### ISSUE TITLE
ObjectPicking Doesn't Work for Multi-Leveled Objects like Human-Body

### ISSUE THREAD

**Issue by jonathanchh1** [Rajawali/Rajawali#2294]
I am building an application where one of the requirements is being able to perform touch on each parts like (finger, head, eyes, knees etc) of a 3D Humanoid Model and briefly modify the texture color. I tried ObjectColorPicking by assigning the each sub-nodes a randomized color however, the ObjectPicking doesn't work. it returns  255, 255, 255 from the glReal byteArray for each object whenever a touch occurs at any coordinates and rgba index returned is -1. I also tried the RayPicking which utilizes BoundingBox or BoundingSphere both for each sub-nodes but all the vectors return x 0, y 0,  z 0 for al positions for transform-min and tranform-max or mMin or mMax.

My Question is, is there away to perform rayCasting or a different technique to handle this use-case with Rajawali?, Thanks in Advance!

**Comment by contriteobserver** [comment]
While it was not specified in the issue, I am assuming that the model is being loaded from a file via a class using our `ILoader` interface. Typically these loaders `parse()` the `Geometry3D` of each child object relative to the origin, which is why `BoundingBox` and `BoundingSphere` always returns the origin.

In `1.2.1970`, we released a new feature in `LoaderOBJ`, one may now call `parse(boolean offsetCentroids)`, when `offsetCentroids` is set to `true`, new behavior causes the `Geometry3D` of the child objects to be referenced to the [centroid](https://en.wikipedia.org/wiki/Centroid) of the child, and for the position of the child to be offset from the origin by that `Vector3D`.

Hopefully this feature will get you there, if you are using an alternative loader, please advise.

**Comment by jonathanchh1** [comment]
Thank you, i just tried it and it seems to detect only one sub-node and ignores the rest of the Nodes and Yes, i'm loading an obj. file.
```
class ChildRayPickingVisitor(private val mRayStart: Vector3, private val mRayEnd: Vector3) : INodeVisitor {
    private val mHitPoint: Vector3 = Vector3()
    var pickedObject: Object3D? = null
        private set

    override fun apply(node: INode) {
        if (node is Object3D) {
            val o = node
            if (!o.isVisible) return
            //RajLog.d("VISITING " + o.getName());
            if (o.geometry.hasBoundingSphere()) {
                val bsphere = o.geometry.boundingSphere
                bsphere.calculateBounds(o.geometry)
                bsphere.transform(o.modelMatrix)
                if (intersectsWith(bsphere)) {
                    if (pickedObject == null ||
                        pickedObject != null && o.position.z < pickedObject!!.position.z
                    ) pickedObject = o
                }
            } else {
                // Assume bounding box if no bounding sphere found.
                val bbox = o.geometry.boundingBox
                bbox.calculateBounds(o.geometry)
                bbox.transform(o.modelMatrix)
                if (intersectsWith(bbox)) {
                    if (pickedObject == null ||
                        pickedObject != null && o.position.z < pickedObject!!.position.z
                    ) pickedObject = o
                }
            }
        }
    }

    private fun intersectsWith(bbox: BoundingBox): Boolean {
        val raySta = mRayStart
        val rayEnd = mRayEnd
        val boxMin = bbox.transformedMin
        val boxMax = bbox.transformedMax

        if (rayEnd.x < boxMin.x && raySta.x < boxMin.x) return false
        if (rayEnd.x > boxMax.x && raySta.x > boxMax.x) return false
        if (rayEnd.y < boxMin.y && raySta.y < boxMin.y) return false
        if (rayEnd.y > boxMax.y && raySta.y > boxMax.y) return false
        if (rayEnd.z < boxMin.z && raySta.z < boxMin.z) return false
        if (rayEnd.z > boxMax.z && raySta.z > boxMax.z) return false
        if (raySta.x > boxMin.x && raySta.x < boxMax.x && raySta.y > boxMin.y && raySta.y < boxMax.y && raySta.z > boxMin.z && raySta.z < boxMax.z) {
            mHitPoint.setAll(raySta)
            return true
        }

        if (getIntersection(
                raySta.x - boxMin.x,
                rayEnd.x - boxMin.x,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.X)
            || getIntersection(
                raySta.y - boxMin.y,
                rayEnd.y - boxMin.y,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.Y)
            || getIntersection(
                raySta.z - boxMin.z,
                rayEnd.z - boxMin.z,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.Z)
            || getIntersection(
                raySta.x - boxMax.x,
                rayEnd.x - boxMax.x,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.X)
            || getIntersection(
                raySta.y - boxMax.y,
                rayEnd.y - boxMax.y,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.Y)
            || getIntersection(
                raySta.z - boxMax.z,
                rayEnd.z - boxMax.z,
                raySta,
                rayEnd
            ) && isInBox(boxMin, boxMax, Vector3.Axis.Z)
        ) {
            Log.d(ChildRayPickingVisitor::class.java.simpleName, "hitPoint=$mHitPoint")
            return true
        }
        return false
    }

    private fun intersectsWith(bsphere: BoundingSphere): Boolean {
        //Log.d(getLocalClassName(), "intersectsWith(BoundingSphere) position=" + bsphere.getPosition().toString() + " r=" + bsphere.getRadius());
        return Intersector.intersectRaySphere(
            mRayStart,
            mRayEnd,
            bsphere.position,
            bsphere.radius,
            mHitPoint
        )
    }

     fun getIntersection(
        fDst1: Double,
        fDst2: Double,
        P1: Vector3,
        P2: Vector3
    ): Boolean {
        if (fDst1 * fDst2 >= 0.0f) return false
        if (floatEqual(fDst1, fDst2)) return false
        mHitPoint.setAll(P1)
        mHitPoint.add(Vector3.subtractAndCreate(P2, P1))
        mHitPoint.multiply(-fDst1 / (fDst2 - fDst1))
        return true
    }

    private fun floatEqual(lhs: Double, rhs: Double): Boolean {
        return Math.abs(lhs - rhs) < .00001f
    }

    private fun isInBox(boxMin: Vector3, boxMax: Vector3, axis: Vector3.Axis): Boolean {
        if (axis == Vector3.Axis.X && mHitPoint.z > boxMin.z && mHitPoint.z < boxMax.z && mHitPoint.y > boxMin.y && mHitPoint.y < boxMax.y) return true
        if (axis == Vector3.Axis.Y && mHitPoint.z > boxMin.z && mHitPoint.z < boxMax.z && mHitPoint.x > boxMin.x && mHitPoint.x < boxMax.x) return true
        return axis == Vector3.Axis.Z && mHitPoint.x > boxMin.x && mHitPoint.x < boxMax.x && mHitPoint.y > boxMin.y && mHitPoint.y < boxMax.y
    }
}
```
is this method accurate? Thank you in advance!

**Comment by contriteobserver** [comment]
Still having trouble reproducing why the `ObjectPicker` interface is not working for body meshes, as one should not have to go the long way round, please clarify what is wrong with this code.
```
    static class ObjectPickingRenderer extends Renderer {
        ObjectColorPicker picker;
        Object3D obj;

        public ObjectPickingRenderer(Context context) {
            super(context);
        }

        @Override
        protected void initScene() {
            picker = new ObjectColorPicker(this);
            picker.setOnObjectPickedListener(new PickedListener());

            try {
                getCurrentScene().setBackgroundColor(Color.CYAN & Color.GRAY);

                DirectionalLight key = new DirectionalLight(-4,-4,-4);
                key.setPower(5/4f);
                getCurrentScene().addLight(key);

                DirectionalLight fill = new DirectionalLight(4,0,0);
                fill.setPower(1/2f);
                getCurrentScene().addLight(fill);

                Material material = new Material();
                material.setColor(Color.MAGENTA & Color.GRAY | Color.DKGRAY);
                material.setDiffuseMethod(new DiffuseMethod.Lambert());
                material.setSpecularMethod(new SpecularMethod.Phong());
                material.setAmbientColor(Color.CYAN | Color.GRAY);
                material.setAmbientIntensity(1/4f,1/4f,1/4f);
                material.enableLighting(true);

                LoaderOBJ loader = new LoaderOBJ(this, R.raw.posing1doll_obj);
                loader.parse();
                obj = loader.getParsedObject();
                registerDecendents(obj);
                getCurrentScene().addChild(obj);

                getCurrentCamera().setUpAxis(Vector3.Axis.Y);
                getCurrentCamera().setPosition(3/2d, 3/2d,2);
                getCurrentCamera().setLookAt(obj.getPosition().clone().add(new Vector3(0,4/3d,0)));
            } catch (Exception e) {
                Log.e(getClass().getSimpleName(), e.getMessage());
            }
        }

        void registerDecendents(Object3D obj) {
            for(int i=0;i<obj.getNumChildren();i++) {
                Object3D child = obj.getChildAt(i);
                if(child.getGeometry().getNumVertices()>0) picker.registerObject(child);
                registerDecendents(child);
            }
        }

        @Override
        public void onOffsetsChanged(float v, float v1, float v2, float v3, int i, int i1) {

        }

        @Override
        public void onTouchEvent(MotionEvent event) {
            picker.getObjectAt(event.getX(), event.getY());
        }
    }

    static class PickedListener implements OnObjectPickedListener {

        @Override
        public void onObjectPicked(@NonNull Object3D object) {
            Log.d(getClass().getSimpleName() + ".onObjectPicked", "picked: " + object.getName());
        }

        @Override
        public void onNoObjectPicked() {
            Log.d(getClass().getSimpleName() + ".onNoObjectPicked", "missed");
        }
    }
```
it seems to pick the children out with this model:
[pose.tar.gz](https://github.com/Rajawali/Rajawali/files/5949974/pose.tar.gz)

**Comment by jonathanchh1** [comment]
Ahh, i see why it wasn't working before, i wasn't calling registerDecendents recursively. I tried it and it definitely works now. Thank you. However, i see that it is changing the Object's name from whatever it was before.. like : **skeleton_left_thumb** name to generatedName **Object1234** etc, do you know how to prevent that because i want to perform animations and object comparison with whichever object is picked? Thank you again in advance!

**Comment by contriteobserver** [comment]
Improving object comparisons sounds like a different issue than `ObjectPicking`.
One issue per issue really helps reduce support overhead, as it makes it easier for others to search through existing answers.

Identifying child objects is a tricky issue, there actually was no "before" object, because we have only implemented one material per object, and because models often have many materials per object, `LoaderOBJ` reconciles this difference by splitting the model into one child object per material. Issue #2301 offers a pull request implementing an approach that might help address some of this problem, `LoaderOBJ` now parses the material name that triggered the creation of a child object. We still need the number portion in the name to ensure uniqueness when one material is shared amongst many objects.

See the fix under issue #2302 for more details.

A no-code workaround is to split the source model into named components before exporting it into an OBJ file, as `LoaderOBJ` has always preserved explicit mesh names.

**Comment by jonathanchh1** [comment]
Oh okay, it makes sense, it should definitely be a different issue and i see you have created threads for them. i'll ask my question concerning that issue over there. Thanks again!
